PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1

```diff
@@ -266,177 +266,181 @@ function Get-PrivilegedUser {
+        # Valid values are "user", "group". If this is not passed, the function will call Graph to dynamically determine the object type.
+        [Parameter()]
+        [string]$Objecttype = "",
+
+        [Parameter()]
+        [int]$Recursioncount = 0
+    )
+
+    # We support group nesting up to 2 levels deep. 
+    # Safeguard: Also protects against infinite loop attacks if there is a circular group assignment in PIM.
+    if ($recursioncount -gt 2) {
+        return
+    }
+
+    if ($Objecttype -eq "") {
+        try {
+            $DirectoryObject = Get-MgBetaDirectoryObject -ErrorAction Stop -DirectoryObjectId $ObjectId
+        } catch {
+            # If the object was probably recently deleted from the directory we ignore it. Otherwise an unhandled 404 causes the tool to crash.
+            if ($_.Exception.Message -match "Request_ResourceNotFound") {
+                Write-Warning "Processing privileged users. Resource $ObjectId may have been recently deleted from the directory because it was not found."
+                return  # Exit the function to ignore this resource and keep the flow going.
             }
+            # If it is a different error, rethrow the error to let the calling function handle it.
+            else {
+                throw $_
+            }
+        }
+
+        $Objecttype = $DirectoryObject.AdditionalProperties."@odata.type" -replace "#microsoft.graph."
+    }
+
+    if ($Objecttype -eq "user") {
+        # If the user's data has not been fetched from graph, go get it
+        if (-Not $PrivilegedUsers.ContainsKey($ObjectId)) {
+            $AADUser = Get-MgBetaUser -ErrorAction Stop -UserId $ObjectId
+            $PrivilegedUsers[$ObjectId] = @{"DisplayName"=$AADUser.DisplayName; "OnPremisesImmutableId"=$AADUser.OnPremisesImmutableId; "roles"=@()}
+Write-Warning "Processing role: $($Role.DisplayName) User: $($AADUser.DisplayName)"
+        }
+        # If the current role has not already been added to the user's roles array then add the role
+        if ($PrivilegedUsers[$ObjectId].roles -notcontains $Role.DisplayName) {
+            $PrivilegedUsers[$ObjectId].roles += $Role.DisplayName
         }
     }
 
-    # Process the Eligible role assignments if the premium license for PIM is there
-    if ($TenantHasPremiumLicense) {
-        # Get a list of all the users and groups that have Eligible assignments
-        $graphArgs = @{
-            "commandlet" = "Get-MgBetaRoleManagementDirectoryRoleEligibilityScheduleInstance"
-            "M365Environment" = $M365Environment }
-        $AllPIMRoleAssignments = Invoke-GraphDirectly @graphArgs
+    elseif ($Objecttype -eq "group") {
+        # In this context $ObjectId is a group identifier
+        $GroupId = $ObjectId
+        # Get all of the group members that are transitively assigned to the current role via group membership
+        $GroupMembers = Get-MgBetaGroupMember -All -ErrorAction Stop -GroupId $GroupId
```

Comment:
suggest capping the number of users in a group that we expand to scan at something reasonable (100? 200?) and printing a warning that users in excess of that are dropped.  At that point they are already well in excess of the policy requirements and we might just end up taking a lot of time crawling the directory for minimal value add.

Reply:
Yeah sure.  I'm mostly concerned with things that can have unbounded performance impact.    I was thinking about 7.1 with the limit.  Good point on not having a limit on 7.2 and 7.3.

I was thinking of fail for global admin and error for the other 2.

But i'm ok with not doing this for now and seeing if it becomes an issue.